azkaban-execserver/src/main/java/azkaban/execapp/event/JobCallbackUtil.java

```diff
@@ -0,0 +1,324 @@
+      String headerPair = headerArray[i];
+      int index = headerPair.indexOf(HEADER_NAME_VALUE_DELIMITER);
+      if (index != -1) {
+        headerList.add(new BasicHeader(headerPair.substring(0, index),
+            headerPair.substring(index + 1)));
+      }
+
+    }
+    return headerList.toArray(new BasicHeader[0]);
+  }
+
+  private static String replaceStatusToken(String template,
+      JobCallbackStatusEnum status) {
+    return template.replaceFirst(STATUS_TOKEN, status.name().toLowerCase());
+  }
+
+  private static StringEntity createStringEntity(String str) {
+    try {
+      return new StringEntity(str);
+    } catch (UnsupportedEncodingException e) {
+      throw new RuntimeException("Encoding not supported", e);
+    }
+  }
+
+  /**
+   * This method takes the job context info. and put the values into a map with
+   * keys as the tokens.
+   * 
+   * @param event
+   * @return Map<String,String>
+   */
+  public static Map<String, String> buildJobContextInfoMap(Event event,
+      String server) {
+
+    if (event.getRunner() instanceof JobRunner) {
+      JobRunner jobRunner = (JobRunner) event.getRunner();
+      ExecutableNode node = jobRunner.getNode();
+      String projectName = node.getParentFlow().getProjectName();
+      String flowName = node.getParentFlow().getFlowId();
+      String executionId =
+          String.valueOf(node.getParentFlow().getExecutionId());
+      String jobId = node.getId();
+
+      Map<String, String> result = new HashMap<String, String>();
+      result.put(CONTEXT_SERVER_TOKEN, server);
+      result.put(CONTEXT_PROJECT_TOKEN, projectName);
+      result.put(CONTEXT_FLOW_TOKEN, flowName);
+      result.put(CONTEXT_EXECUTION_ID_TOKEN, executionId);
+      result.put(CONTEXT_JOB_TOKEN, jobId);
+      result.put(CONTEXT_JOB_STATUS_TOKEN, node.getStatus().name().toLowerCase());
+
+      /*
+       * if (node.getStatus() == Status.SUCCEEDED || node.getStatus() ==
+       * Status.FAILED) { result.put(JOB_STATUS_TOKEN,
+       * node.getStatus().name().toLowerCase()); } else if (node.getStatus() ==
+       * Status.PREPARING) { result.put(JOB_STATUS_TOKEN, "started"); }
+       */
+
+      return result;
+
+    } else {
+      throw new IllegalArgumentException("Provided event is not a job event");
+    }
+  }
+
+  /**
+   * Replace the supported tokens in the URL with values in the contextInfo.
+   * This will also make sure the values are HTTP encoded.
+   * 
+   * @param value
+   * @param contextInfo
+   * @param withEncoding - whether the token values will be HTTP encoded
+   * @return String - value with tokens replaced with values
+   */
+  public static String replaceTokens(String value,
+      Map<String, String> contextInfo, boolean withEncoding) {
+
+    String result = value;
+    String tokenValue =
+        encodeQueryParam(contextInfo.get(CONTEXT_SERVER_TOKEN), withEncoding);
+    result = result.replaceFirst(Pattern.quote(CONTEXT_SERVER_TOKEN), tokenValue);
```

Comment:
Why use replaceFirst rather than replaceAll?
Documents do not mention it at all.